sov-modules/sov-app-template/src/lib.rs

```diff
@@ -84,24 +105,46 @@ where
         _misbehavior_hint: Option<Self::MisbehaviorProof>,
     ) -> Result<
         Vec<Vec<sovereign_sdk::stf::Event>>,
+        // Question: Should we return an enum error here? (see TODOs bellow)
         sovereign_sdk::stf::ConsensusSetUpdate<OpaqueAddress>,
     > {
+        let mut batch_workspace = WorkingSet::new(self.current_storage.clone());
+        batch_workspace = batch_workspace.to_revertable();
+
+        // TODO: Error handling.
+        match self
+            .tx_hooks
+            .enter_apply_batch(sequencer, &mut batch_workspace)
+        {
+            Ok(_) => {}
+            Err(_) => todo!(),
+        }
+
         let mut events = Vec::new();
 
         // Run the stateless verification.
-        let txs = self
+        let txs = match self
             .tx_verifier
             .verify_txs_stateless(batch.take_transactions())
-            .or(Err(ConsensusSetUpdate::slashing(sequencer)))?;
-        let mut batch_workspace = WorkingSet::new(self.current_storage.clone());
+        {
+            Ok(txs) => txs,
+            Err(_) => {
+                self.revert_and_slash(batch_workspace);
+                return Err(ConsensusSetUpdate::slashing(sequencer));
+            }
+        };
 
         for tx in txs {
             batch_workspace = batch_workspace.to_revertable();
             // Run the stateful verification, possibly modifies the state.
-            let verified_tx = self
-                .tx_hooks
-                .pre_dispatch_tx_hook(tx, &mut batch_workspace)
-                .or(Err(ConsensusSetUpdate::slashing(sequencer)))?;
+            let verified_tx = match self.tx_hooks.pre_dispatch_tx_hook(tx, &mut batch_workspace) {
+                Ok(verified_tx) => verified_tx,
+                Err(_) => {
+                    // TODO: should we revert here? // pre_dispatch_tx_hook changes nonces
```

Comment:
Yeah, I think revert is the right thing here - a malicious bundle should have no effect except for slashing the sequencer (so that we can short-circuit by jumping to the invalid transaction in some cases)